docs/containers/debug-python.md

```diff
@@ -0,0 +1,149 @@
+---
+Area: containers
+ContentId: f9ffec31-9253-4f71-a4eb-79ea7b3a8f55
+PageTitle: Configure and troubleshoot debugging of Python apps running in a Docker container
+DateApproved: 01/29/2020
+MetaDescription: How to configure and troubleshoot debugging of Python apps running in a Docker container, using Visual Studio Code.
+---
+
+# Debug Python within a container
+
+When adding Docker files to a Python project, tasks and launch configurations are added to enable debugging the application within a Docker container. To accommodate the various scenarios of Python projects, some apps may require additional configuration.
+
+## Configuring the Docker container entry point
+
+The Docker extension infers the entry point of the Docker container via properties of `tasks.json`. This container entry point is automatically configured when using the **Docker: Add Docker Files to Workspace...** command.
+
+### Example: Configuring the entry point for a Python module
+
+```json
+{
+  "tasks": [
+    {
+      "type": "docker-run",
+      "label": "docker-run: debug",
+      "dependsOn": [
+        "docker-build"
+      ],
+      "python": {
+        "args": [
+          "runserver",
+          "0.0.0.0:8000",
+          "--nothreading",
+          "--noreload"
+        ],
+        "module": "manage"
+      }
+    }
+  ]
+}
+```
+
+### Example: Configuring the entry point for a Python file
+
+```json
+{
+  "tasks": [
+    {
+      "type": "docker-run",
+      "label": "docker-run: debug",
+      "dependsOn": [
+        "docker-build"
+      ],
+      "python": {
+        "file": "hello_app/webapp.py"
+      }
+    }
+  ]
+}
+```
+
+## Automatically launching the browser to the entry page of the application
+
+When the **Docker: Python - Django** or **Docker: Python - Flask** launch configurations are selected, the Docker extension will automatically launch the browser to the main page of the app. While this feature is enabled by default, this behavior can be configured explicitly via the `dockerServerReadyAction` object in `launch.json`.
+
+This feature depends on several aspects of the application:
+
+- The application **must output to the debug console or Docker logs**.
+- The application must log a "server ready" message.
+- The application must serve a browsable page.
+
+Here is an example of using `dockerServerReadyAction` to launch the browser to open the `about.html` page based on a specific server message pattern:
+
+```json
+{
+  "configurations": [
+    {
+      "name": "Docker: Python - Django",
+      "type": "docker",
+      "request": "launch",
+      "preLaunchTask": "docker-run: debug",
+      "python": {
+        "pathMappings": [
+          {
+            "localRoot": "${workspaceFolder}",
+            "remoteRoot": "/app"
+          }
+        ],
+        "projectType": "django"
+      },
+      "dockerServerReadyAction": {
+        "action": "openExternally",
+        "pattern": "Starting development server at (https?://\\S+|[0-9]+)",
```

Comment:
That regular expression might need a brief explanation ?

Reply:
Tagging @haniamr for an of how the pattern & uriFormat interact. I'll update it with a ">**Note**: " below the json example

Reply:
The expression simply attempts to capture: "Starting development server at http://localhost:8000". It accommodates variations in the url such http or https, any host name and port.

The relation between the pattern and uriFormat is as follows:
- We capture a group from the regex (i.e. anything enclosed in (...)), so in the above example it will be the url
- If the matched group is a port number, then only 1 placeholder (i.e. %s) in the uriFormat is expected
- If the matched group is a url, then 2 placeholders are expected, one for the scheme and the other for the port

Hope that helps!

Reply:
So would there be an error here if the log output message only said "Starting development server at 8000"? Because in this case it would only be capturing the port and not the scheme?

Reply:
I added Hani's explanation.